ecosystem/sep-0039.md

```diff
@@ -0,0 +1,196 @@
+
+Notice that ownership of the token directly conveys ownership of a unit of the asset, entirely within the Stellar ledger. Other scenarios such as locking down the issuing account to avoid supply increases, ownership transfers via payments, etc. are built on top of this basic model and are thus extraneous to this SEP.
+
+Note also that this SEP does not address concerns around *immutability*. While there is an optional checksum component (addressed later), the account itself is free to modify the data at-will if it is not locked.
+
+Our exclusive focus in this SEP is Step 2: **storing the asset within the account.**
+
+### Data Format
+We begin by describing the raw data format, then dive into each component:
+
+  - index (2 bytes)
+  - version (1 byte)
+  - metadata length, `n` (varying)
+  - metadata (`n` bytes)
+  - binary data (varying)
+
+For example,
+
+    00|1|26|text/plain;l=13,image/webp|TPwJh>}A"=r@@Y?Fetc...
+
+(Here, the `|` characters are semantic, included just to separate the pieces in the list above; they are not actually in the data itself.)
+
+The following sections break down each component in detail.
+
+#### Index
+Since an account can only have 1000 entries, we only need 10 bits for an index field. We _only_ need indices present so that the data entries are ASCII-sortable in Horizon. In fact, decoders can ignore it entirely if they wish. Thus, the index is expressed as a "base 36" value: 0-9 then a-z.
+
+A valid SEP-39 entry always begins with `00`, indicating that this is the 0th row of data. In fact, every row's key _must_ begin with a two-character, [Base36] index.
+
+---------
+
+**Author's note**: This portion is definitely up for debate. Base36 (well, more accurately, a radix of 36) allows us to shave a byte off of each row (saving up to 1kB total), but we could stick to a 000-999 model (stringified integer) for simplicity. We could also do `aa` through `zz`, because as noted, the indices are _purely_ for sorting. 
+
+Note that supporting a 36-radix is trivial in any language, since we know the value is strictly an integer. For example,
+
+```python
+import string 
+ALPHABET = string.digits + string.ascii_lowercase
+
+def int_to_b36(i: int) -> str:
+    d, r = divmod(i, len(ALPHABET))
+    return ALPHABET[d] + ALPHABET[r]
+
+def b36_to_int(s: str) -> int:
+    return ALPHABET.index(s[0]) * len(ALPHABET) + ALPHABET.index(s[1])
+```
+
+---------
+
+#### Version
+The first entry's index is followed by the version string. For the current version of this SEP, this is just `1`.
+
+#### Metadata
+The **metadata** is a comma-separated list of [Media Type]s alongside any associated parameters.
+
+It is prefixed by a **length**, which is just a stringified integer (that is, a 32-byte metadata will have "32" as its length). When parsing, you should consume characters until hitting a non-digit, since a valid [Media Type] will never start with a digit. Note that row indices are not included in the metadata length.
```

Comment:
>When parsing, you should consume characters until hitting a non-digit

This seems brittle to me, and it may be difficult to parse such a message safely, or at the least it might be easy to pass it in a way that is unsafe. e.g. If you read forever because someone sends you a long stream of numbers. I think we should follow in the footsteps of other message formats, which is ultimately what we're defining here, and either define a fixed number of bytes that the length fits into, or define a clear signal for how long the length should be. One way to do this is to specify a length-length, or another is to use a variable integer format where 7 bits of integer value are encoded into every byte with the high-bit indicating there's another byte, such as how protocol buffers encode integers.

I think we should lean heavily towards this being simple and easy to implement, but on the same note I see the desire to want to optimize this as much as possible. We could use XDR to define this format. That would make for increased safety, but won't necessarily be efficient since XDR pads out to multiples of 4 and all lengths are 4 bytes. 

The simpler thing, I think, would be to encode the length as binary uint16.

Reply:
I'll defer to @Shaptic on this one but if we do this do we also need to do anything with the `l=` param? Or will the fact that it's encapsulated in metadata context keep it from key:value overflow issues? We could also just cap the length to the max possible bytes. Anything beyond that should be considered an error. So (62+64)×1,000=126,000. That would be the max integer so I think we can leave as is with the caveat for interpreters to fail if the meta length or any individual media type is > 126000.

Reply:
We can't use protobufs nor xdr nor int16 because this will be in the key, and thus printable ASCII @leighmcculloch. That's part of the overall challenge, unfortunately. I do agree that it's brittle, though...

I think it's reasonable to say that the length (a) _must_ fit in the first key and (b) _must_ be <= 126000. The question then becomes: should we always have 6 characters for metadata length, e.g. `000014`, or should we say "up to 6 digits"? We could also use hex, so `00000` -> `1ec30` are the valid lengths.

Thankfully, the efficiency sacrifice is not on a per-row basis, so the implementation-ease vs. efficiency trade-off is more reasonable. Thoughts?

> do we also need to do anything with the `l=` param?

I think because it's not relevant until the entire metadata + buffer is parsed, it's okay. Obviously someone can have phony lengths, but this is natural to defend against.

Reply:
I vote leave it as is. Mixing different int parsers get's weird. Use our custom parser for the index, plain for the version, hex for the metadata length, plain again for the metadata lengths. I'd prefer to leave it as the custom for index, plain for everything else with the understanding that the metadata length (a) must fit in the first key and (b) must be <= 126000.

Reply:
Yep, agreed on all counts :+1: